gh-109523: Reading text from a non-blocking stream with read may now raise a BlockingIOError if the operation cannot immediately return bytes.

giosiragusa opened the pull request:
- Updated `_io_TextIOWrapper_read_impl` to raise a `BlockingIOError` when `read()` does not return bytes.
- Introduced `test_read_non_blocking` to ensure that a `BlockingIOError` is raised when attempting to read from a non-blocking pipe with no data using `os.pipe()`.

<!-- gh-issue-number: gh-109523 -->
* Issue: gh-109523
<!-- /gh-issue-number -->

giosiragusa committed:
Add unittest to handle non-blocking read scenarios.

- Introduced `test_read_non_blocking` to ensure proper handling of non-blocking reads using `os.pipe()`.
- The test verifies that a `BlockingIOError` is raised when attempting to read from a non-blocking pipe with no data.
- Updated code to raise `BlockingIOError` with a specific message when `read()` does not return bytes.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

cmaloney commented on Modules/_io/textio.c:
```diff
@@ -1992,6 +1992,12 @@ _io_TextIOWrapper_read_impl(textio *self, Py_ssize_t n)
             goto fail;
 
+        if (bytes == Py_None){
+            PyErr_SetString(PyExc_BlockingIOError, "read() should return bytes");
```
I think the exception text here is too much about library internals (behavior of `read` at different levels of the I/O stack), rather than guiding the user to the documentation and/or saying what the likely causes are (NonBlocking I/O, maybe a underlying I/O stack bug since a user could potentially swap out RawIO/FileIO)

giosiragusa replied:
Thank you for the feedback! I've updated the exception message to be more informative and user-friendly. The new message now reads:

```c
PyErr_SetString(PyExc_BlockingIOError, "Unexpected None encountered. This may be due to non-blocking I/O "
                                       "or an issue with the underlying I/O implementation. Please refer to "
                                       "the documentation or ensure that the I/O operation is properly "
                                       "configured.");
```

I hope this provides clearer guidance for users encountering this issue.

cmaloney commented on Lib/test/test_io.py:
```diff
@@ -3919,6 +3919,22 @@ def test_issue35928(self):
+        r, w = os.pipe()
+        try:
+            os.set_blocking(r, False)
+            with open(r, 'r') as textfile:
```
TextIOWrapperTest that this is in (https://github.com/python/cpython/blob/main/Lib/test/test_io.py#L2720) is subclassed so it can be run under the two main I/O implementations in python ([`_io`](https://github.com/python/cpython/blob/main/Lib/test/test_io.py#L3942), [`_pyio`](https://github.com/python/cpython/blob/main/Lib/test/test_io.py#L4042)). Here, need to use `self.io.open` to get the implementation-specific `open()` being tested.

For clarity I think it would also be good to do `open(r, 'rt')` rather than relying on "text is the default" since the test is specifically trying to test TextIO behavior

giosiragusa replied:
Thank you for the insightful suggestion! I've made the recommended changes to ensure that the test runs under both I/O implementations and to clarify the mode being used. The updated code now reads:

```python
with self.io.open(r, 'rt') as textfile:
    r = None
    # Nothing has been written so a non-blocking read raises a BlockingIOError exception.
    with self.assertRaises(BlockingIOError):
```

This should improve clarity and correctness in the context of testing `TextIOWrapper`.

cmaloney reviewed (changes requested):
The code in `_pyio`'s implementation of TextIO should ideally be changed in this PR as well so the two main implementations stay in sync (I think, in specific the code around https://github.com/python/cpython/blob/main/Lib/_pyio.py#L2288-L2293, when updating the test to use `self.io.open` the test I _think_ will fail on PyIO which should help validate where it should go to match behavior)

Definitely liking the direction, looking forward to the next iteration

giosiragusa committed:
Update documentation for Text I/O and Buffered Streams to include handling of BlockingIOError in non-blocking I/O scenarios

giosiragusa committed:
Fix double backticks for inline literals.

giosiragusa committed:
Add a more meaningful message for the BlockingIOError.

giosiragusa committed:
Updated the TextIOWrapper test to use self.io.open() instead of open() to ensure the test runs under both _io and _pyio implementations. Also, explicitly specified 'rt' mode in the open() call for clarity, as the test is focused on TextIO behavior.

giosiragusa committed:
- Raise `BlockingIOError` in `TextIOWrapper.read()` in `_pyio.py` when `buffer.read()` returns zero bytes.
- Fix double backticks in the documentation.

cmaloney commented on Doc/library/io.rst:
```diff
@@ -64,6 +64,13 @@ In-memory text streams are also available as :class:`StringIO` objects::
+.. note::
+   If you are working with a non-blocking stream, be aware that operations on text I/O
+   objects may raise a ``BlockingIOError``. This occurs when the underlying stream is
+   in non-blocking mode and a read operation cannot be completed immediately, potentially
```
I think the `, potentially leading to a ``BlockingIOError``.` part can be dropped in this sentence / is redundant with the "may raise" part of the sentence earlier on.

giosiragusa replied:
Thank you for your feedback and suggestions. I found them very helpful, especially since I don't have daily practice with writing standard library documentation. I've made changes based on your recommendations and addressed all the points you raised. I hope these adjustments align better with the standard library style now.

cmaloney commented on Doc/library/io.rst:
```diff
@@ -770,6 +777,11 @@ than raw I/O does.
+      .. note::
+         When the underlying raw stream is non-blocking, a ``BlockingIOError``
+         may be raised if the read operation cannot be completed immediately.
+         Ensure proper exception handling in such cases.
```
Personal nitpick / my individual style leanings: I don't think the "Ensure proper exception handling in such cases" part is needed. Definitely not a python doc style expert though / leave to others.

cmaloney commented on Lib/_pyio.py:
```diff
@@ -2523,9 +2523,14 @@ def read(self, size=None):
         decoder = self._decoder or self._get_decoder()
         if size < 0:
+
+            b = self.buffer.read()
```
personal nitpick / nit: I like longer than one character variable names

giosiragusa replied:
I usually prefer longer variable names as well! I was using `b` as a shorthand for `bytes` to avoid shadowing the global variable. 😅 Since I was quite tired at the time, I didn't fully consider the naming. With a fresh perspective, I’ve now chosen a more suitable name. I hope this works better!

cmaloney commented on Modules/_io/textio.c:
```diff
@@ -1992,6 +1992,15 @@ _io_TextIOWrapper_read_impl(textio *self, Py_ssize_t n)
 
+        if (bytes == Py_None){
+            PyErr_SetString(PyExc_BlockingIOError, "Unexpected None encountered. This may be due to non-blocking I/O "
+                                                   "or an issue with the underlying I/O implementation. Please refer to "
```
personal nitpick / nit: I don't think the "Please refer to" part here I don't think is needed stylistically, but also don't know the python documentation standards really well

cmaloney commented on Misc/NEWS.d/next/C_API/2024-08-12-10-15-19.gh-issue-109523.S2c3fi.rst:
```diff
@@ -0,0 +1 @@
+In _io_TextIOWrapper_read_impl, raise BlockingIOError when read() from a non-blocking pipe does not return bytes.
```
This updates `_pyio` as well now, Maybe something like "Reading text from a non-blocking stream with `.read()` may now throw a NonBlockingIO error when Non-blocking I/O is encountered."

giosiragusa replied:
It took a bit of trial and error to get Sphinx-lint to accept the new text, but I think it’s finally in good shape.

cmaloney reviewed (approved):
Overall I think all the code is correct. I have some personal stylistic nitpicks (How I program / what I'd match doesn't match what you did, but I don't have any strong argument to leave or change, take them or leave them as you wish). Think this needs a more experienced Python reviewer than me at this point. 

Thanks a lot for the improvements!

giosiragusa committed:
Update NEWS entry to include references to _io and _pyio for BlockingIOError change

Revised the NEWS entry in `MISC/NEWS.d/next/C_API` to specify that the recent 
change affecting the `BlockingIOError` now applies to both `_io.TextIOWrapper.read()` 
and `_pyio.TextIOWrapper.read()`.

giosiragusa committed:
Align documentation and code with Python style practices

- Adapted documentation to reflect common Python style practices.
- Refactored code to use a clearer variable name.

giosiragusa committed:
Fix news message for sphinx-lint

giosiragusa committed:
Fix news message for sphinx-lint

giosiragusa committed:
Fix news message for sphinx-lint

giosiragusa committed:
Make notes more consistent

z764969689 commented on Lib/_pyio.py:
```diff
@@ -2523,9 +2523,14 @@ def read(self, size=None):
+
+            chunk = self.buffer.read()
+            if chunk is None:
+                raise BlockingIOError
```
I found the error message under the C implementation is clear while for py it's empty and could be confusing as well. What about keeping the same error message to ensure the consistency between c & py?

giosiragusa replied:
You raise a valid point. I'm not sure how I missed that in the Python implementation, but I've addressed it now and aligned the error message with the C version for consistency.

giosiragusa committed:
Align BlockingIOError message in _pyio.py with the C implementation in textio.c

picnixz commented on Doc/library/io.rst:
```diff
@@ -64,6 +64,12 @@ In-memory text streams are also available as :class:`StringIO` objects::
    f = io.StringIO("some initial text data")
 
+.. note::
+   When working with a non-blocking stream, be aware that operations on text I/O objects
```
For notes, you can have an empty line between the directive and the text.

giosiragusa replied:
Done.

picnixz commented on Lib/_pyio.py:
```diff
@@ -2523,9 +2523,15 @@ def read(self, size=None):
                 size = size_index()
         decoder = self._decoder or self._get_decoder()
         if size < 0:
+
```
Can we remove unnecessary blank lines?

giosiragusa replied:
Done.

picnixz commented on Lib/_pyio.py:
```diff
@@ -2523,9 +2523,15 @@ def read(self, size=None):
+            if chunk is None:
+                raise BlockingIOError("Unexpected None encountered. This may be due to non-blocking I/O or an issue "
+                                      "with the underlying I/O implementation.")
+
```
Ditto

giosiragusa replied:
Done.

gpshead commented on Modules/_io/textio.c:
```diff
@@ -1992,6 +1992,13 @@ _io_TextIOWrapper_read_impl(textio *self, Py_ssize_t n)
+        if (bytes == Py_None){
+            PyErr_SetString(PyExc_BlockingIOError, "Unexpected None encountered. This may be due to non-blocking I/O "
+                                                   "or an issue with the underlying I/O implementation.");
+            return NULL;
```
Py_DECREF(bytes); first.

giosiragusa replied:
Done.

picnixz commented on Lib/_pyio.py:
```diff
@@ -2523,9 +2523,15 @@ def read(self, size=None):
+
+            chunk = self.buffer.read()
+            if chunk is None:
+                raise BlockingIOError("Unexpected None encountered. This may be due to non-blocking I/O or an issue "
```
I'm not sure we want to say that it's an issue with the underlying implementation. Just say that nothing could be read.

giosiragusa replied:
I followed the suggestion of @gpshead and kept the message simple.

gpshead commented on Modules/_io/textio.c:
```diff
@@ -1992,6 +1992,13 @@ _io_TextIOWrapper_read_impl(textio *self, Py_ssize_t n)
             goto fail;
 
+        if (bytes == Py_None){
+            PyErr_SetString(PyExc_BlockingIOError, "Unexpected None encountered. This may be due to non-blocking I/O "
```
Lets keep the error message simple (in both versions), "read returned None".

giosiragusa replied:
Done.

gpshead commented on Doc/library/io.rst:
```diff
@@ -64,6 +64,12 @@ In-memory text streams are also available as :class:`StringIO` objects::
    f = io.StringIO("some initial text data")
 
+.. note::
+   When working with a non-blocking stream, be aware that operations on text I/O objects
```
"reads" or "read operations" rather than "operations".   and change the later text to "if the stream cannot perform the operation immediately."

giosiragusa replied:
Done.

giosiragusa committed:
Add blank line after note.

giosiragusa committed:
Remove unnecessary blank lines.

giosiragusa committed:
Keep the error message simple.

giosiragusa committed:
Release the reference before throwing the exception.

giosiragusa committed:
Add blank line after notes.

giosiragusa committed:
Remove unnecessary blank lines.

giosiragusa committed:
Added what's new entry.
Added entry in ACKS.

picnixz commented on Doc/whatsnew/3.14.rst:
```diff
@@ -117,6 +117,12 @@ Added support for converting any objects that have the
+___
+
+* Reading text from a non-blocking stream with ``read`` may now raise a :exc:`BlockingIOError` if the operation cannot immediately return bytes.
+(Contributed by Giovanni Siragusa in :gh:`109523`.)
```
```suggestion

io
--

* Reading text from a non-blocking stream with ``read`` may now raise a
  :exc:`BlockingIOError` if the operation cannot immediately return bytes.
  (Contributed by Giovanni Siragusa in :gh:`109523`.)

```

picnixz replied:
FTR: we want two blank lines before the section "io" (each module is separated by two blank lines, but the contributions within a same module are separated by 1 blank line in general).

giosiragusa committed:
Update Doc/whatsnew/3.14.rst

Fix missing indents and other syntax issues.

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>
